[JSC] Switch from Shewchuk’s Algorithm to xsum Algorithm to Optimize `Math.sumPrecise`

Gumichocopengin8 opened the pull request:
#### 2682d2ab17a40851f7e6bcff35126e810b0646fd
<pre>
[JSC] Switch from Shewchuk’s Algorithm to xsum Algorithm to Optimize `Math.sumPrecise`
<a href="https://bugs.webkit.org/show_bug.cgi?id=293158">https://bugs.webkit.org/show_bug.cgi?id=293158</a>

Reviewed by Yusuke Suzuki.

Previously, Shewchuk&apos;s algorithm was used [1], but it became slow with large arrays.
Improving the performance of Math.sumPrecise is important because it may be used in
scientific calculations, charting, and statistical analysis of large data sets,
where both high precision and fast speed are essential.
This patch switches to Radford M. Neal&apos;s xsum algorithm [2],
which improves the performance of Math.sumPrecise.
The original implementation is in C [3]; a C++ translation was referenced for integration [4].

                                TipOfTree              Patched                Ratio
math-sumPrecise-10              3.4638+-0.0681         2.6122+-0.0922         1.2338x faster
math-sumPrecise-100             27.2897+-0.115         17.9712+-0.0483        1.5185x faster
math-sumPrecise-1000            283.2722+-4.3743       171.4865+-2.4889       1.6518x faster
math-sumPrecise-10000           2922.8715+-9.1500      1709.5876+-4.2845      1.7096x faster

[1]: <a href="https://github.com/WebKit/WebKit/commit/f49d6a41fb8d93f9c7e6758d87a06a9eea3e9698">https://github.com/WebKit/WebKit/commit/f49d6a41fb8d93f9c7e6758d87a06a9eea3e9698</a>
[2]: <a href="https://arxiv.org/abs/1505.05571">https://arxiv.org/abs/1505.05571</a>
[3]: <a href="https://gitlab.com/radfordneal/xsum">https://gitlab.com/radfordneal/xsum</a>
[4]: <a href="https://github.com/Gumichocopengin8/xsum.cpp">https://github.com/Gumichocopengin8/xsum.cpp</a>

* JSTests/microbenchmarks/math-sumPrecise-10.js: Added.
* JSTests/microbenchmarks/math-sumPrecise-100.js: Added.
* JSTests/microbenchmarks/math-sumPrecise-1000.js: Added.
* JSTests/microbenchmarks/math-sumPrecise-10000.js: Added.
* Source/WTF/wtf/PreciseSum.cpp:
(WTF::XsumSmallAccumulator::XsumSmallAccumulator):
(WTF::PreciseSum::PreciseSum):
(WTF::PreciseSum::addList):
(WTF::PreciseSum::add):
(WTF::PreciseSum::compute):
(WTF::PreciseSum::xsumSmallAddInfNan):
(WTF::PreciseSum::xsumAdd1NoCarry):
(WTF::PreciseSum::xsumCarryPropagate):
(WTF::PreciseSum::incrementWhenValueAdded):
(WTF::twosum): Deleted.
* Source/WTF/wtf/PreciseSum.h:

Canonical link: <a href="https://commits.webkit.org/295242@main">https://commits.webkit.org/295242@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4021da2cae8422231f852b94a32d2df8394566fa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14192 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109364 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54836 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32416 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79122 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107174 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18830 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93974 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59449 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18642 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12013 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54196 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96843 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88349 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111750 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102779 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23096 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88145 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31688 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90162 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87805 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32690 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10442 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25788 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16962 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31253 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36566 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/126413 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31047 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34955 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34383 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32607 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/PreciseSum.cpp:
```diff
@@ -59,123 +65,358 @@
-    double hi = x + y;
-    double lo = y - (hi - x);
-    return std::make_pair(hi, lo);
+    size_t offset = 01;
```
Is this `01` right? If it is, then use `1`.

Gumichocopengin8 replied:
it should be `0`. Thanks for reviewing code and finding a typo. 
So far `addList` function isn't used, but planning to use it in a different PR for speed improvement.

Gumichocopengin8 committed:
[JSC] Switch from Shewchuk’s Algorithm to xsum Algorithm to Optimize `Math.sumPrecise`
https://bugs.webkit.org/show_bug.cgi?id=293158

Reviewed by Yusuke Suzuki.

Previously, Shewchuk's algorithm was used [1], but it became slow with large arrays.
Improving the performance of Math.sumPrecise is important because it may be used in
scientific calculations, charting, and statistical analysis of large data sets,
where both high precision and fast speed are essential.
This patch switches to Radford M. Neal's xsum algorithm [2],
which improves the performance of Math.sumPrecise.
The original implementation is in C [3]; a C++ translation was referenced for integration [4].

                                TipOfTree              Patched                Ratio
math-sumPrecise-10              3.4638+-0.0681         2.6122+-0.0922         1.2338x faster
math-sumPrecise-100             27.2897+-0.115         17.9712+-0.0483        1.5185x faster
math-sumPrecise-1000            283.2722+-4.3743       171.4865+-2.4889       1.6518x faster
math-sumPrecise-10000           2922.8715+-9.1500      1709.5876+-4.2845      1.7096x faster

[1]: https://github.com/WebKit/WebKit/commit/f49d6a41fb8d93f9c7e6758d87a06a9eea3e9698
[2]: https://arxiv.org/abs/1505.05571
[3]: https://gitlab.com/radfordneal/xsum
[4]: https://github.com/Gumichocopengin8/xsum.cpp

* JSTests/microbenchmarks/math-sumPrecise-10.js: Added.
* JSTests/microbenchmarks/math-sumPrecise-100.js: Added.
* JSTests/microbenchmarks/math-sumPrecise-1000.js: Added.
* JSTests/microbenchmarks/math-sumPrecise-10000.js: Added.
* Source/WTF/wtf/PreciseSum.cpp:
(WTF::XsumSmallAccumulator::XsumSmallAccumulator):
(WTF::PreciseSum::PreciseSum):
(WTF::PreciseSum::addList):
(WTF::PreciseSum::add):
(WTF::PreciseSum::compute):
(WTF::PreciseSum::xsumSmallAddInfNan):
(WTF::PreciseSum::xsumAdd1NoCarry):
(WTF::PreciseSum::xsumCarryPropagate):
(WTF::PreciseSum::incrementWhenValueAdded):
(WTF::twosum): Deleted.
* Source/WTF/wtf/PreciseSum.h:

Canonical link: https://commits.webkit.org/295242@main
